NanoUI 3.0!

neersighted opened the pull request:
A complete rewrite of the NanoUI frontend to be more modern, and yet still run on potatoes (XP/IE8).

Note that this makes breaking changes to template syntax, and introduces new build tools.

This resolves #13460.
This resolves #13641.

Also, snuck in a stealth-change, minimaps are now stored per-map, meaning we don't have to regenerate them on map change.

![Demo](http://i.imgur.com/QtNyZD8.png)

neersighted committed:
Begin rewrite of NanoUI frontend.

* Remove jQuery in favor of polyfills.
* Make NanoUI work with IE8-IE11
* Upstream doT
* Change NanoUI template syntax
* Event driven code

neersighted committed:
Port NanoUI styles to LESS

neersighted commented:
Thanks to @tkdrg for testing on Windows XP/IE8 and WINE.

optimumtact commented:
looks good to me

WJohn commented:
30k lines of code removed and 105 file changes? Did you screw something up or were there really that many changes, because that's insane.

neersighted commented:
Yep, it really changes/removes that much.

Dorsisdwarf commented:
There's a lot of whitespace and CSS being removed, which boosts the removal count

Jalleo commented:
@neersighted does it recognise if the map has been updated sincelast run or not?

neersighted commented:
Indeed it does, using md5sums.

duncathan commented:
>30,000 deletions

YOU BEAUTIFUL BEAUTIFUL PERSON

ike709 commented:
>+1,333 −30,528 for a presumably better system

And the Coder of the Day award goes to...

optimumtact commented:
guys they just removed all the scripts and made them load from the vendor provided cdn's

Lo6a4evskiy commented:
Yeah, lines of code is a bad way to measure anything.

neersighted committed:
Restructure NanoUI frontend; use build tools

Yes, you now need Node.js... I'm sorry, but this makes
it so much easier to write NanoUI.

We now precompile all the things...

Dorsisdwarf commented:
It removes 30,000 lines of code that have to be compiled and downloaded and shit

Shadowlight213 commented:
Neersighted for best coder 2015

neersighted committed:
Coffeescript; Bower; Gulp; asset_cache

We now write frontend code in Coffeescript, alongside our LESS.
PostCSS adds hacks to make CSS work on old versions of IE.
Bower is used to manage dependencies.
Gulp is used to build the code.

Images are dropped in favor of SVG or CSS effects.

NanoUI is now a 'simple' asset in the asset_cache

Dorsisdwarf commented:
soah many commit

Shadowlight213 commented:
@neersighted yeah, you're probably going to want to squash your commits at some point.

neersighted commented:
That is the plan... This is still very much WIP. I don't usually shitcommit :p

neersighted commented:
PR updated with preview of new Air Alarm.

neersighted committed:
Add NanoUI README

neersighted committed:
Modularize CSS and JS; SVG

Purge NanoUI assets as we now have a gorgeous Nanotrasen SVG logo
Thanks to @YotaXP for helping me with cleaning up the logo

Razharas commented:
Ask for live testing on servers again

neersighted commented:
Redesigned some stuff, here's what it looks like now:

![Demo](http://i.imgur.com/QtNyZD8.png)

neersighted committed:
Prefix everything with 'nanoui' to avoid conflicts

neersighted committed:
Rework how NanoUI handles Topic()

Also, add the ability to winset and the like, all from Javascript

neersighted committed:
Enable NanoUI to go chromeless

Add a new 'Fancy NanoUI' preference to enable chromeless mode

bear1ake commented:
It's happening?

optimumtact commented:
Soon

neersighted commented:
I think this is now ready to testmerge. If the testmerge goes well I will squash my commits and be ready to merge.

KorPhaeron commented:
We can not test merge this because it involves preferences and will mess up save files

At least I was told as much for hand of god

KorPhaeron commented on code/modules/power/apc.dm:
```diff
@@ -771,81 +771,72 @@
+		if("overload")
+			if(usr.has_unlimited_silicon_privilege)
+				src.overload_lighting()
+		if("hack")
```
This is the option I want to show up for traitor AI's

Do you know where I go to change what shows up on the list?

neersighted replied:
You need to edit the malf status proc to make malf options show up for traitor AIs.

neersighted committed:
Re-organize nanoui.dm

YotaXP committed:
Re-write JSON writer to output valid JSON

How the old one worked, I don't know...
We don't have a reader yet, but we don't use a reader sooo...

neersighted committed:
Add a loading spinner to NanoUI...

This is kinda dirty because we use external resources; however, the
point is that if assets are not loaded, we need to show a spinner.

neersighted committed:
Improve focus code; fix SVG

Replace the NT SVG with a better version
Always try to give focus back to the mapwindow

Also make links more obvious

neersighted committed:
Make NanoUI resistant to Topic spoofs

Move Topic() into a NanoUI-specific ui_act proc
Update to @YotaXP's latest JSON code.
Return focus to the mapwindow if a key is pressed in a NanoUI.
